sys/ztimer: add 'ztimer_no_periph_rtt

fjmolinas opened the pull request:
### Contribution description

This is a temporary fix for Issue #17060. It allows disabling
auto inclusion of `ztimer_periph_rt*` in cases where another
module or application requires direct access.

Limitations:
- as ifeq are involved order of inclusion matters, therefore
  these modules should be included early in the build at the application
  level and not in modules `Makefile.dep`
- this does not disallow direct inclusions of `ztimer_periph_rt*`,
  since this only disables auto inclusion of these modules

This is a temporary solution since this is already possible with
Kconfig, but not in make.

This implements the easiest fix proposed in #17060 to be able to move forward with the migration.

### Testing procedure

```
USEMODULE="ztimer_msec ztimer_no_periph_rtt" BOARD=samr21-xpro make -C tests/ztimer_msg/ info-modules  | grep periph_rtt
ztimer_no_periph_rtt
```
```
DISABLE_MODULES=ztimer_periph_rtt_auto_select BOARD=samr21-xpro make -C tests/ztimer_msg/ info-modules  | grep periph_rtt
```

### Issues/PRs references

#17060

kfessel commented:
Why not just remove the automatic rtt rtc selection from makefile.dep?

periph_rtt is no good indicator for an rtt being able to power msec
periph_rtc does conflict with rtt on many devices

why did this change get merged with an KConfig PR #16909?

fjmolinas commented:
> periph_rtt is no good indicator for an rtt being able to power msec

Why not?

> periph_rtc does conflict with rtt on many devices

There isl` FEATURES_CONFLICTING` for that

> why did this change get merged with an KConfig PR #16909?

You are surely right about that, I should have asked for a split, at the time I though it was an issue with the Makefile not being inline with Kconfig, but thas was my bad.

> just revert 34259e1 and remove the Makefile.dep change in c80390f

I won't revert the change in `Makefile.dep` but I'll provide a removal of the default rtc seleciton.

benpicco commented on sys/include/ztimer.h:
```diff
@@ -197,6 +197,11 @@
  *
+ * If `periph_rtt` or `periph_rtc` is required with direct access by another
+ * MODULE or an application, `ztimer_no_periph_rtt` or `ztimer_no_periph_rtc`
+ * can be included to avoid automatic selection of either `ztimer_periph_rtc`
```
```suggestion
 * must be included to avoid automatic selection of either `ztimer_periph_rtc`
```

kfessel reviewed (changes requested):
waiting for #17113
after that ztimer will not default to rtc 

This no longer needs a Makefile.dep split

fjmolinas commented:
@benpicco also mentione that a build-system warning would be desirable, I'll take another look on the next sprint

fjmolinas commented:
@benpicco I added a warning in the `periph/rtt.h` header, in the `ztimer.h` header, and when building if both modules are included a warning is shown. Anything else you would suggest?

I also added in d336016ad4e6ffd6b7656ec53409d7f39902558c an alternative based on `DEFAULT_MODULE`

benpicco commented:
The `DEFAULT_MODULE` solution sounds ideal, then modules that make use of the RTT API can just `DISABLE_MODULE += ztimer_periph_rtt`

fjmolinas commented:
> The `DEFAULT_MODULE` solution sounds ideal, then modules that make use of the RTT API can just `DISABLE_MODULE += ztimer_periph_rtt`

No, this is not the case. Rember that `DEFAULT_MODULE` only works if they are disabled before other dependencies are parsed, otherwise they will already be selected. It would need to be disabled in something included in `makefiles/defaultmodules.inc.mk` or in the application. `DEFAULT_MODULE` can not be disabled during dependency resolution because at that stage they were already converted to a `USEMODULE`

kfessel commented on sys/ztimer/Makefile.dep:
```diff
@@ -89,33 +89,10 @@ endif
-  else
-    USEMODULE += ztimer_periph_timer
-  endif
+  include $(RIOTBASE)/sys/ztimer/Makefile.ztimer.rtt.dep
```
```suggestion
  ifneq (,$(filter ztimer_periph_rtt_auto_select,$(USEMODULE)))
  FEATURES_OPTIONAL += periph_rtt
  # HACK: periph_rtt will get used only in the next iteration but an updated
  # state for FEATURES_USED is needed here so include `features_check.inc.mk`
  # here instead.
  # An other option would be to check FEATURES_PROVIDED this would avoid the
  # order of inclusion problem but it would no take into account possible conflicts
  # and is also currently not allowed in the build system.
  # An other alternative would be to delay to the next loop, but this produce a
  # case where another loop is not executed and the conditional not evaluated
  # If these kind of usecases pop up before Kconfig migration is completed
  # then another alternative would be introduce a variable to require an extra
  # loop independent of USEMODULE, FEATURES_REQUIRED and USEPKG
  include $(RIOTMAKE)/features_check.inc.mk
  ifneq (,$(filter periph_rtt,$(FEATURES_USED)))
    USEMODULE += ztimer_periph_rtt
  else
    USEMODULE += ztimer_periph_timer
  endif
  endif
```

kfessel commented on sys/ztimer/Makefile.dep:
```diff
@@ -89,33 +89,10 @@ endif
 
 ifneq (,$(filter ztimer_msec,$(USEMODULE)))
```
```suggestion
ifneq (,$(filter ztimer_msec,$(USEMODULE))$(filter ztimer_sec,$(USEMODULE)))
```

kfessel commented on sys/ztimer/Makefile.dep:
```diff
@@ -89,33 +89,10 @@ endif
-  else
-    USEMODULE += ztimer_periph_timer
-  endif
+  include $(RIOTBASE)/sys/ztimer/Makefile.ztimer.rtt.dep
```
```suggestion
```

moved above

kfessel commented on sys/ztimer/Makefile.ztimer.rtt.dep:
```diff
@@ -0,0 +1,20 @@
+ifneq (,$(filter ztimer_periph_rtt_auto_select,$(USEMODULE)))
+  FEATURES_OPTIONAL += periph_rtt
+  # HACK: periph_rtt will get used only in the next iteration but an updated
+  # state for FEATURES_USED is needed here so include `features_check.inc.mk`
+  # here instead.
+  # An other option would be to check FEATURES_PROVIDED this would avoid the
+  # order of inclusion problem but it would no take into account possible conflicts
+  # and is also currently not allowed in the build system.
+  # An other alternative would be to delay to the next loop, but this produce a
+  # case where another loop is not executed and the conditional not evaluated
+  # If these kind of usecases pop up before Kconfig migration is completed
+  # then another alternative would be introduce a variable to require an extra
+  # loop independent of USEMODULE, FEATURES_REQUIRED and USEPKG
+  include $(RIOTMAKE)/features_check.inc.mk
+  ifneq (,$(filter periph_rtt,$(FEATURES_USED)))
+    USEMODULE += ztimer_periph_rtt
+  else
+    USEMODULE += ztimer_periph_timer
+  endif
+endif
```
```suggestion
```

kfessel reviewed (commented):
no need for `sys/ztimer/Makefile.ztimer.rtt.dep`

benpicco commented on sys/ztimer/Makefile.include:
```diff
@@ -4,6 +4,8 @@ ifneq (,$(filter ztimer_xtimer_compat,$(USEMODULE)))
   PSEUDOMODULES += xtimer
 endif
 
+MODULES_ZTIMER_ON_RTT_CONFLICT = rtt_rtc
```
```suggestion
MODULES_ZTIMER_ON_RTT_CONFLICT = rtt_rtc gnrc_lwmac,
```

benpicco commented:
So what's missing here?

kfessel commented:
I think the `sys/ztimer/Makefile.ztimer.rtt.dep` can be part of `sys/ztimer/Makefile.dep` other than that it just needs to be squashed

fjmolinas commented:
> I think the `sys/ztimer/Makefile.ztimer.rtt.dep` can be part of `sys/ztimer/Makefile.dep` other than that it just needs to be squashed

Should I go with the `DEAFULT_MODULE` version then? This is now blocking #17122 since unittests use a mock rtt...

kfessel commented:
Yes,  `DEFAULT_MODULE` seems the right way to go
I still think: 
> sys/ztimer/Makefile.ztimer.rtt.dep can be part of sys/ztimer/Makefile.dep
other than that

please squash

kfessel commented on sys/ztimer/Kconfig:
```diff
@@ -31,6 +31,12 @@ config MODULE_ZTIMER_PERIPH_RTT
     depends on HAS_PERIPH_RTT
     select MODULE_PERIPH_RTT
 
+config MODULE_ZTIMER_NO_PERIPH_RTC
+    bool "Do not auto include RTC peripheral"
+
```
this can be removed (since rtc is never be auto selected)

kfessel commented on sys/ztimer/Kconfig:
```diff
@@ -83,12 +89,12 @@ config ZTIMER_SEC_BACKEND_TIMER
     bool "RTT"
     depends on HAS_PERIPH_RTT
-    select MODULE_ZTIMER_PERIPH_RTT
+    select MODULE_ZTIMER_PERIPH_RTT if !MODULE_ZTIMER_NO_PERIPH_RTT
 
 config ZTIMER_SEC_BACKEND_RTC
     bool "RTC"
     depends on HAS_PERIPH_RTC
-    select MODULE_ZTIMER_PERIPH_RTC
+    select MODULE_ZTIMER_PERIPH_RTC if !MODULE_ZTIMER_NO_PERIPH_RTC
 
```
maybe this is not the way it is in the dep (but as you know i am not good at reading kconfig)

Makefiles are so much easier << this is wrong

kfessel reviewed (commented):
sorry didn't see that the (seems to not have followed the .dep changes)
please ignore and ping me if i misread.

fjmolinas commented:
Sorry for the delay, squashed.

kfessel commented on drivers/include/periph/rtt.h:
```diff
@@ -22,6 +22,11 @@
  * that it is ensured that the RTT will function properly while it is enabled.
  *
+ * @warning     This module will be automatically be used as a backend for @ref
+ *              `ztimer_msec` and @ref `ztimer_sec`. If direct acces to RTT is needed
```
```suggestion
 *              `ztimer_msec` and @ref `ztimer_sec`. If direct access to RTT is needed
```
codespell

kfessel commented on drivers/include/periph/rtt.h:
```diff
@@ -22,6 +22,11 @@
  * On many CPUs, certain power states might need to be blocked in rtt_init(), so
```
doccheck has issues with this line:
i am not sure if it is about this file: if so please fix, 
but i think it is about an other file (generic doxygen noise increase) (where rtt_init can be found) (in that case do not fix in this PR))

kfessel reviewed (commented):
one (two) minor static-test-failures.

fjmolinas commented:
I'll squash if static tests pass.

kfessel commented on sys/ztimer/Makefile.dep:
```diff
@@ -8,6 +8,10 @@ ifneq (,$(filter ztimer,$(USEMODULE)))
 
+  # If ztimer is used enable rtt autoselection for ztimer_msec, and ztimer_sec.
+  # This will enable inclusion of `ztimer_periph_rtt` if `periph_rtt` is supported.
+  DEFAULT_MODULE += ztimer_periph_rtt_auto_select
```
this line is not yet evaluated when l 96 is reached -> no auto selection is done

fjmolinas commented on sys/ztimer/Makefile.dep:
```diff
@@ -8,6 +8,10 @@ ifneq (,$(filter ztimer,$(USEMODULE)))
 
+  # If ztimer is used enable rtt autoselection for ztimer_msec, and ztimer_sec.
+  # This will enable inclusion of `ztimer_periph_rtt` if `periph_rtt` is supported.
+  DEFAULT_MODULE += ztimer_periph_rtt_auto_select
```
Ajj true, that is why initially I had put in Makefile.default..., couldn't remember why and moved it out :/

fjmolinas replied:
I can't think of another solution than ae0bf90e7935e954908c7ea1705315c004f3dd4c... it does mean that a `ztimer` module will show up even when `ztimer` is not used...

fjmolinas commented:
Rebased because of conflict with master

leandrolanzieri commented on sys/include/ztimer.h:
```diff
@@ -197,6 +197,11 @@
+ * If `periph_rtt` or is required with direct access by another MODULE or
+ * application, `ztimer_periph_rtt_auto_select` can be disabled to avoid automatic
+ * selection of `ztimer_periph_rtt` as a backend for ZTIMER_SEC and ZTIMER_MSEC.
+ * i.e.: `DISABLE_MODULE += ztimer_periph_rtt_auto_select`.
```
Could we maybe invert the logic? Making a module that avoids the default selection if present in the application makefile? then we wouldn't have the issue of the default module always showing up, right?

fjmolinas replied:
That was the initial approach I had, I can go back to that.

fjmolinas replied:
@kfessel thoughts?

kfessel replied:
i think the was a problem that lead to abondoning that approach (like could not be kconfig model???)

fjmolinas replied:
I split into an alternative the old versio so its easier to compare https://github.com/RIOT-OS/RIOT/pull/17284

fjmolinas replied:
@leandrolanzieri @kfessel opinion on what option to prefer this one or https://github.com/RIOT-OS/RIOT/pull/17284?

kfessel reviewed (commented):
while reviewing these line threw me in the wrong way would you please remove the `S` in the make variables (even though these lines are not part of this PR)

```
sys/ztimer/Makefile.dep
12:  # but RIOT's auto_init scheme expects "auto_init_ztimer" in DEFAULT_MODULES so
13:  # it can be disabled (by adding to DISABLE_MODULES).
please change to 
DEFAULT_MODULE
DISABLE_MODULE
```

fjmolinas commented on sys/include/ztimer.h:
```diff
@@ -197,6 +197,11 @@
  *              if if these are missing it will use same basic timer
  *              as ZTIMER_USEC does.
  *
+ * If `periph_rtt` or is required with direct access by another MODULE or
```
```suggestion
 * If `periph_rtt` is required with direct access by another MODULE or
```

fjmolinas commented:
> sys/ztimer/Makefile.dep
12:  # but RIOT's auto_init scheme expects "auto_init_ztimer" in DEFAULT_MODULES so
13:  # it can be disabled (by adding to DISABLE_MODULES).
please change to 
DEFAULT_MODULE
DISABLE_MODULE

This was removed in another PR.

fjmolinas commented:
Superseeded by #17284
